File: src/usr_msg_topic_example/type_coverage_talker/src/type_coverage_talker.cc
/******************************************************************************
 * Copyright (c) 2022-2026 SEGAR
 * SPDX-License-Identifier: Apache-2.0
 *****************************************************************************/

#include <cstdint>
#include <memory>

#include "test_msgs/msg/TypeCoverage.hpp"

#include "segar/segar.h"
#include "segar/time/clock.h"

namespace {

using TypeCoverage = test_msgs::msg::TypeCoverage;

constexpr char kTopic[] = "/topic/usr_msg/type_coverage";

std::shared_ptr<TypeCoverage> MakeTypeCoverage(uint32_t seq) {
  auto msg = std::make_shared<TypeCoverage>();
  msg->ready((seq % 2) == 0);
  msg->int8_value(static_cast<int8_t>(seq));
  msg->uint8_value(static_cast<uint8_t>(seq));
  msg->int16_value(static_cast<int16_t>(seq));
  msg->uint16_value(static_cast<uint16_t>(seq));
  msg->int32_value(static_cast<int32_t>(seq));
  msg->uint32_value(seq);
  msg->int64_value(static_cast<int64_t>(seq));
  msg->uint64_value(rti::segar::Time::Now().ToNanosecond());
  msg->temperature(25.0f + static_cast<float>(seq) * 0.1f);
  msg->voltage(3.3 + static_cast<double>(seq) * 0.05);
  msg->note("sample-" + std::to_string(seq));

  msg->raw_bytes().assign({0xA0, 0xB0, 0xC0});
  msg->py_bytes().assign({0xA0, 0xB0, 0xC0});
  msg->py_bytearray().assign({static_cast<int8_t>(-5), static_cast<int8_t>(-4),
                              static_cast<int8_t>(-3)});
  msg->py_memoryview().assign({0xD0, 0xD1, 0xD2});
  msg->fixed_bytes()[0] = 1;
  msg->fixed_bytes()[1] = 2;
  msg->fixed_bytes()[2] = 3;
  for (size_t i = 0; i < msg->samples().size(); ++i) {
    msg->samples()[i] = static_cast<int32_t>(seq * 10 + i);
  }
  msg->detail().detail_id(static_cast<int32_t>(seq));
  msg->detail().description("detail-" + std::to_string(seq));
  return msg;
}

}  // namespace

int main(int argc, char* argv[]) {
  (void)argc;
  RETURN_VAL_IF(!rti::segar::Init(argv[0]), EXIT_FAILURE);
  auto node = rti::segar::CreateNode("usr_msg_type_coverage_talker");
  RETURN_VAL_IF(!node, EXIT_FAILURE);

  auto writer = node->CreateWriter<TypeCoverage>(kTopic);
  RETURN_VAL_IF(!writer, EXIT_FAILURE);

  uint32_t seq = 1;
  auto callback = [&writer, &seq]() {
    auto msg = MakeTypeCoverage(seq++);
    AINFO_IF(!writer->Write(msg)) << "Failed to write TypeCoverage";
    AINFO << "Sent TypeCoverage int32=" << msg->int32_value()
          << ", note=" << msg->note();
  };

  auto timer = std::make_shared<rti::segar::Timer>(1000, callback, false);
  timer->Start();
  rti::segar::WaitForShutdown();
  return EXIT_SUCCESS;
}

File: src_python/usr_msg_topic_example/type_coverage_talker/src/type_coverage_talker.py
#!/usr/bin/env python3

import time

from segar.python import segar
from test_msgs.msg import TypeCoverage, TypeCoverageDetail


TOPIC = "/topic/usr_msg/type_coverage"


def main():
    if not segar.Init("py_usr_msg_type_coverage_talker"):
        return 1
    node = segar.CreateNode("py_usr_msg_type_coverage_talker")
    if node is None:
        return 1
    writer = node.CreateWriter(TOPIC, TypeCoverage)
    if writer is None:
        print("[py_usr_msg_type_coverage_talker] create writer failed")
        return 1

    seq = 1
    while segar.OK():
        msg = TypeCoverage()
        msg.ready = seq % 2 == 0
        msg.int8_value = seq
        msg.uint8_value = seq
        msg.int16_value = seq
        msg.uint16_value = seq
        msg.int32_value = seq
        msg.uint32_value = seq
        msg.int64_value = seq
        msg.uint64_value = int(time.time_ns())
        msg.temperature = 25.0 + seq * 0.1
        msg.voltage = 3.3 + seq * 0.05
        msg.note = f"sample-{seq}"
        msg.raw_bytes = bytes([0xA0, 0xB0, 0xC0])
        msg.py_bytes = bytes([0xA0, 0xB0, 0xC0])
        msg.py_bytearray = bytearray([0xFB, 0xFC, 0xFD])
        msg.py_memoryview = memoryview(bytearray([0xD0, 0xD1, 0xD2]))
        msg.fixed_bytes = bytes([1, 2, 3])
        msg.samples = [seq * 10 + i for i in range(12)]
        msg.detail = TypeCoverageDetail(detail_id=seq, description=f"detail-{seq}")
        if writer.Write(msg):
            print(
                "[py_usr_msg_type_coverage_talker] sent: "
                f"int32={msg.int32_value}, py_memoryview={list(msg.py_memoryview)}"
            )
        else:
            print("[py_usr_msg_type_coverage_talker] write failed")
        seq += 1
        time.sleep(1.0)

    segar.WaitForShutdown()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
